File: node-ocr/src/extract.ts
import { ExtractConfig, BaseStructuredOutput, validateConfig } from './types';
import { processDocExtractionMistral } from './providers/mistral';
import { processDocExtractionOpenAI } from './providers/openai';
import { processDocExtractionGemini } from './providers/gemini';
import { parseDocumentLocal } from './providers/local';

/**
 * Extracts structured information from a document using AI providers or local parsing
 * @param config Extraction configuration including provider, file path, API key, and response format schema
 * @returns Structured data according to the provided schema
 */
export async function extract<T extends BaseStructuredOutput>(config: ExtractConfig): Promise<T> {
  const provider = config.provider || 'mistral';

  // Local provider doesn't need API key validation
  if (provider !== 'local') {
    validateConfig(config);
  }

  switch (provider) {
    case 'mistral': {
      const result = await processDocExtractionMistral(config);
      return result.parsed as T;
    }
    
    case 'openai': {
      const result = await processDocExtractionOpenAI(config);
      return result.parsed as T;
    }
    
    case 'gemini': {
      const result = await processDocExtractionGemini(config);
      return result.parsed as T;
    }
    
    case 'local': {
      const result = await parseDocumentLocal(config.filePath);
      return result as unknown as T;
    }
    
    default:
      throw new Error(`Provider ${provider} not supported: ${provider}`);
  }
}

File: python-ocr/src/docuglean/extract.py
"""
Document extraction module for Docuglean Python SDK.
"""


from .providers.gemini import process_doc_extraction_gemini
from .providers.huggingface import process_doc_extraction_huggingface
from .providers.mistral import process_doc_extraction_mistral
from .providers.openai import process_doc_extraction_openai
from .types import ExtractConfig, StructuredExtractionResult, validate_config


async def extract(config: ExtractConfig) -> StructuredExtractionResult:
    """
    Extracts structured information from a document using specified provider.

    Args:
        config: Extraction configuration including provider, file path, API key, and response format schema

    Returns:
        Structured data according to the provided schema

    Raises:
        ValueError: If configuration is invalid
        Exception: If provider is not supported
    """
    # Default to mistral if no provider specified
    provider = config.provider or "mistral"

    # Validate configuration
    validate_config(config)

    # Route to correct provider
    if provider == "mistral":
        return await process_doc_extraction_mistral(config)
    elif provider == "openai":
        return await process_doc_extraction_openai(config)
    elif provider == "huggingface":
        return await process_doc_extraction_huggingface(config)
    elif provider == "gemini":
        return await process_doc_extraction_gemini(config)
    else:
        raise Exception(f"Provider {provider} not supported yet")
